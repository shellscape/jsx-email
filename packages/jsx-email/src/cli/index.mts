#!/usr/bin/env node --enable-source-maps --no-warnings=ExperimentalWarning

import './main.mjs';
