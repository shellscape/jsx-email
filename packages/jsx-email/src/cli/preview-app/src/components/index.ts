export * from './code';
export * from './heading';
export * from './icons';
export * from './logo';
export * from './sidebar';
export * from './text';
export * from './topbar';
