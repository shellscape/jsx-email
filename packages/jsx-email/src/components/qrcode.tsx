import * as React from 'react';

import qrcode from 'qrcode';

type RootProps = React.ComponentPropsWithoutRef<'img'>;

export interface QrProps extends RootProps {
  correctionLevel: 'L' | 'M' | 'H';
  height?: number;
  size: number;
  style?: React.CSSProperties;
  width?: number;
}

export const QrCode: React.FC<Readonly<QrProps>> = ({
  correctionLevel,
  alt,
  src,
  width,
  height,
  style,
  size,
  ...props
}) => {
  const qrCodeOptions = {
    errorCorrectionLevel: correctionLevel,
    height: size,
    margin: 1,
    type: 'image/png',
    width: size
  };
  const qrCodeBuffer = qrcode.toBuffer(src, qrCodeOptions);

  return (
    <img
      {...props}
      data-id="jsx-email/qr"
      alt={alt}
      src={qrCodeBuffer}
      width={width && width < size ? size : width || size}
      height={height && height < size ? size : height || size}
      style={{
        ...style
      }}
    />
  );
};
