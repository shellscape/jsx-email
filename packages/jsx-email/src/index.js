import './helpers.js';
// components
export * from './components/background.js';
export * from './components/body.js';
export * from './components/button.js';
export * from './components/butan.js';
export * from './components/code.js';
export * from './components/color-scheme.js';
export * from './components/column.js';
export * from './components/conditional.js';
export * from './components/container.js';
export * from './components/font.js';
export * from './components/graph.js';
export * from './components/head.js';
export * from './components/heading.js';
export * from './components/hr.js';
export * from './components/html.js';
export * from './components/img.js';
export * from './components/link.js';
export * from './components/markdown.js';
export * from './components/preview.js';
export * from './components/row.js';
export * from './components/raw.js';
export * from './components/section.js';
export * from './components/tailwind/tailwind.js';
export * from './components/text.js';
// renderer
export * from './renderer/compat/context.js';
export * from './renderer/compat/hooks.js';
export * from './renderer/jsx-to-string.js';
export * from './renderer/render.js';
export { useData } from './renderer/suspense.js';
export * from './plugins.js';
export * from './types.js';
//# sourceMappingURL=index.js.map