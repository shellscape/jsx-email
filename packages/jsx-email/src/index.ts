export * from './components/body';
export * from './components/button/button';
export * from './components/code';
export * from './components/color-scheme';
export * from './components/column';
export * from './components/conditional';
export * from './components/container';
export * from './components/font';
export * from './components/head';
export * from './components/heading';
export * from './components/hr';
export * from './components/html';
export * from './components/img';
export * from './components/link';
export * from './components/markdown';
export * from './components/preview';
export * from './components/row';
export * from './components/section';
export * from './components/tailwind/tailwind';
export * from './components/text';
export * from './render';
export * from './types';
