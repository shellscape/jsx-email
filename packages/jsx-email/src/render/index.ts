export * from './render';
export * from './jsx-to-string';
export * from './process';
