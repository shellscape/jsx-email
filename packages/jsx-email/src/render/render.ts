import { htmlToText } from 'html-to-text';

import type { PlainTextOptions, RenderOptions } from '../types';

import { jsxToString } from './jsx-to-string';
import { processHtml } from './process';

export const renderPlainText = async (
  component: React.ReactElement,
  options?: PlainTextOptions
) => {
  const result = await jsxToString(component);
  return htmlToText(result, {
    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-skip="true"]' },
      { options: { linkBrackets: false }, selector: 'a' }
    ],
    ...options
  });
};

export const render = async (component: React.ReactElement, options?: RenderOptions) => {
  const { minify = false, plainText, pretty = false } = options || {};

  if (plainText) return renderPlainText(component, typeof plainText === 'object' ? plainText : {});

  let html = await jsxToString(component);

  html = await processHtml({ html, minify, pretty });

  return html;
};
