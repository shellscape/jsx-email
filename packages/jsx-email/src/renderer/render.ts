import { htmlToText } from 'html-to-text';

import { defineConfig, loadConfig, mergeConfig, type JsxEmailConfig } from '../config.js';
import { callHook, callProcessHook } from '../plugins.js';
import type { PlainTextOptions, RenderOptions } from '../types.js';

import { jsxToString } from './jsx-to-string.js';
import { getMovePlugin } from './move-style.js';
import { unescapeForRawComponent } from './raw.js';

export const jsxEmailTags = ['jsx-email-cond'];

export const renderPlainText = async (
  component: React.ReactElement,
  options?: PlainTextOptions
) => {
  const { formatters, selectors } = options || {};

  const result = await jsxToString(component);
  return htmlToText(result, {
    formatters: {
      raw: (elem, _walk, builder) => {
        if (elem.children.length && elem.children[0].type === 'comment') {
          builder.addInline(unescapeForRawComponent(elem.children[0].data!.trim()));
        }
      },
      ...formatters
    },

    selectors: [
      { format: 'skip', selector: 'img' },
      { format: 'skip', selector: '[data-skip="true"]' },
      { options: { linkBrackets: false }, selector: 'a' },
      {
        format: 'raw',
        options: {},
        selector: 'jsx-email-raw'
      },
      ...(selectors || [])
    ],
    ...options
  });
};

export const render = async (component: React.ReactElement, options?: RenderOptions) => {
  let config = await loadConfig();
  if (config.render.plainText || options?.plainText)
    return renderPlainText(
      component,
      typeof options?.plainText === 'object' ? options.plainText : {}
    );

  const renderOptions = { render: options };

  if (options) {
    // Note: structuredClone chokes on symbols
    const { symbol: _, ...cloneTarget } = config as any;
    const merged = await mergeConfig(cloneTarget, renderOptions);

    config = await defineConfig(merged);
  }

  let html = await jsxToString(component);

  html = await callHook({ config, hookType: 'beforeRender', html });
  html = await processHtml(config, html);
  html = await callHook({ config, hookType: 'afterRender', html });
  return html;
};

const processHtml = async (config: JsxEmailConfig, html: string) => {
  const { rehype } = await import('rehype');
  const { default: stringify } = await import('rehype-stringify');
  const docType =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';
  const movePlugin = await getMovePlugin();
  const settings = { emitParseErrors: true };
  const reJsxTags = new RegExp(`<[/]?(${jsxEmailTags.join('|')})>`, 'g');

  // @ts-ignore: This is perfectly valid, see here: https://www.npmjs.com/package/rehype#examples
  const processor = rehype().data('settings', settings);

  processor.use(movePlugin);
  await callProcessHook({ config, processor });

  const doc = await processor
    .use(stringify, {
      allowDangerousCharacters: true,
      allowDangerousHtml: true,
      closeEmptyElements: true,
      collapseEmptyAttributes: true
    })
    .process(html);

  let result = docType + String(doc).replace('<!doctype html>', '').replace('<head></head>', '');

  result = result.replace(reJsxTags, '');
  result = result.replace(/<jsx-email-raw.*?><!--(.*?)--><\/jsx-email-raw>/g, (_, p1) =>
    unescapeForRawComponent(p1)
  );

  return result;
};
