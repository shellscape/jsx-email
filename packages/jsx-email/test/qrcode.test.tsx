// @ts-ignore
import React from 'react';

import { jsxToString } from '../src/render';
import { QrCode } from '../src/components/qrcode';

describe('render', async () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  it('renders QR code with required props', async () => {
    const testValue = 'test';
    const html = await jsxToString(<QrCode src={testValue} size={300} correctionLevel="L" />);
    expect(html).toContain('width="300"');
    expect(html).toContain('height="300"');
  });

  it('passes style and other props correctly', async () => {
    const style = { border: '1px solid black' };
    const html = await jsxToString(
      <QrCode src="test" size={300} correctionLevel="L" style={style} data-testid="qr-test" />
    );
    expect(html).toContain('style="border:1px solid black"');
    expect(html).toContain('data-testid="qr-test"');
  });

  it('renders the <QrCode> component', async () => {
    const actualOutput = await jsxToString(
      <QrCode src="Lorem ipsum" size={300} correctionLevel="L" />
    );
    expect(actualOutput).toMatchSnapshot();
  });
});
