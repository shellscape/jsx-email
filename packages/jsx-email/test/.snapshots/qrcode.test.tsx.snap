// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render > renders the <QrCode> component 1`] = `"<img data-id=\\"jsx-email/qr\\" width=\\"300\\" height=\\"300\\" style=\\"\\"/>"`;
